### ISSUE TITLE
Assertion failure in BotBaseMovementAction

### ISSUE THREAD

**Issue by viciious** [Qfusion/qfusion#408]
"BotBaseMovementAction::Assert(): An assertion has failed"

@dnk777 Appears randomly in rekt gt. Can't provide any further information because stack appears either corrupted or impossible to unwind.

**Comment by dnk777** [comment]
I am acknowledged and is looking at it

**Comment by viciious** [comment]
On a side note, bots revert to standing still on rekt quite a lot, completely ignoring the player.

**Comment by dnk777** [comment]
I have spotted the source of a failure easily, sorry for that. As a temporary workaround to make the game playable I suggest commenting this line: https://github.com/Qfusion/qfusion/blob/master/source/game/ai/bot_movement.cpp#L4958. The full patch comes ASAP.

> On a side note, bots revert to standing still on rekt quite a lot, completely ignoring the player.

Tell what the map is exactly please.

**Comment by dnk777** [comment]
> On a side note, bots revert to standing still on rekt quite a lot, completely ignoring the player.

This might come from that I use a slightly modified BSPC now, I should find and post a link ASAP.

**Comment by viciious** [comment]
> Tell what the map is exactly please.

wda5

**Comment by dnk777** [comment]
I have just tested rekt gametype on a map wda5 and they hunt and attack me a lot, roaming over the map.
Try using https://github.com/Warsow/bspc/commits/qf-player-hull please.

**Comment by viciious** [comment]
Commenting out the assertion results in another assertion failure:
> Game Error: t=000026963 OriginLazyVar::Value(): Attempt to get a value of var #2 which is not in PRESENT state

**Comment by dnk777** [comment]
I have been testing the branch in its initial state (diverged in August 2016) for a week on a dedicated server in order to prepare weights, though I must admit i have never compiled it without PUBLIC_BUILD define (the console output without it is spammy and even leads to nohup/wswconsole logs overflow in a 24 hours for >20 bots). And also I have been testing rekt too (as I know you would run it first) and have manually created weights for it.

> Commenting out the assertion results in another assertion failure:

I either have to setup an MSVC environment and it could take about a day, or replace G_Error() call
https://github.com/Qfusion/qfusion/blob/master/source/game/ai/ai.cpp#L90
with something that can produce a stack trace and post it here please.

**Comment by viciious** [comment]
![image](https://user-images.githubusercontent.com/1173058/32665574-d1700cda-c645-11e7-9c3e-5bf337dc7748.png)

**Comment by viciious** [comment]
Recompiling the aas file helped mildly, bots tend to stand still sometimes, unless they spot me.

**Comment by dnk777** [comment]
Thanks, and it seems worse than i thought because the condition above forbids an access to this var.

https://github.com/Qfusion/qfusion/blob/master/source/game/ai/world_state.h#L896
Try replacing this line by 
`AI_FailWith( "OriginLazyVar::Value()", "Attempt to get a value of var #%hd which is not in PRESENT state, var state bits are %d\n", index, (int)StateBits() );`

(Stack trace is not needed in this case)

**Comment by viciious** [comment]
> Game Error: t=000025991 OriginLazyVar::Value(): Attempt to get a value of var #2 which is not in PRESENT state, var state bits are 12

**Comment by dnk777** [comment]
Thats my fault, sorry
https://github.com/Qfusion/qfusion/blame/master/source/game/ai/world_state.h#L371
I am going to hack it in our appveyor builds first.

**Comment by dnk777** [comment]
Another movement issue is spotted, is going to be fixed, and I propose temporarily changing this condition 
https://github.com/Qfusion/qfusion/blob/master/source/game/ai/bot_movement.cpp#L4464
as a quick fix until a movement patch arrives to
`if( suggestedLookDirs.size() >= maxSuggestedLookDirs ) {`

**Comment by viciious** [comment]
Btw,

>2>d:\projects\qfusion\source\game\ai\bot_movement.h(1440): warning C4018: '<': signed/unsigned mismatch
2>d:\projects\qfusion\source\game\ai\bot_movement.h(1445): warning C4018: '<': signed/unsigned mismatch

**Comment by viciious** [comment]
Didn't help with the original assertion:

![image](https://user-images.githubusercontent.com/1173058/32670984-bb53b1e8-c656-11e7-9c2d-3f0e50df7dea.png)

**Comment by dnk777** [comment]
I have not submitted a movement patch yet, that was a GOAP fix. Thats why I have proposed using these quick fixes posted in this thread temporarily.

**Comment by viciious** [comment]
Just hit another assertion on wda5:

> msg = 0x012ec010 "Game Error: t=000060579 BotLandOnSavedAreasMovementAction::FindJumppadAreaNum(): Can't find an AAS area num for the jumppad @ 272.0 120.0 -60.0\n"

**Comment by dnk777** [comment]
>msg = 0x012ec010 "Game Error: t=000060579 BotLandOnSavedAreasMovementAction::FindJumppadAreaNum(): Can't find an AAS area num for the jumppad @ 272.0 120.0 -60.0\n"

This is weird as this assertion has never been guarded by any defines remaining in public release builds, and i have never met triggering it. This is the jumppad leading to the glass dome. I have seen bots using it without triggering any assertions, and I run a debug + no-public + address sanitizer build.

Also I have pasted this 
https://gist.github.com/dnk777/24e1ffbd2e96ac3503db667e6540b3e9
 in place of ai/ai.cpp/AI_CommonFrame(). This code follows the `BotLandOnSavedAreasMovementAction::FindJumppadAreaNum()` 
code almost literally, AAS world data/trigger data does not get modified after loading, and this code never fails (i have tried various maps, the only condition is having a valid AAS file).

**Comment by viciious** [comment]
Ok, the assertions now seem to have been fixed. However, bots still act retarded in dm mode on wda5, I'll post the AAS file, hold on.

**Comment by viciious** [comment]
To reproduce, load wda5, callvote deathmatch, spawn 1 bot and ready up. Observe the bot eventually abandon all maneuvers and then suicide.
[wda5.zip](https://github.com/Qfusion/qfusion/files/1462700/wda5.zip)

**Comment by dnk777** [comment]
Do not pass "optimize" flag to the BSPC, it is important (it just strips very important face/edge info).

**Comment by dnk777** [comment]
The AAS file size for wda5 should be around 650 Kb

**Comment by viciious** [comment]
This is an unoptimized version. BSPC is known to produce different AAS files depending on the arch (x86 vs x64).

**Comment by dnk777** [comment]
I am observing the moronic behavior you have described using the file you have given, but not using my own produced AAS files.

**Comment by viciious** [comment]
May be a bspc bug then. I'm clearly using the branch you pointed me to, compiled on x86 with MSVC

**Comment by dnk777** [comment]
If it produces different files depending of an arch, it is fundamentally broken. I have run DM with a single bot on wda5 for 5 minutes and the bot has never suicided, roaming over the map and grabbing items all the time (yes the bot rarely attacked me if i have not attacked it first, thats what DM weights are, if you want a bot rush headless and attack players, just copy rekt.weights to dm_wda5.weights in base/ai).

**Comment by dnk777** [comment]
Files are intended to be same for different arch. At least they should be compiled on a single arch chosen once and memory representation should be decoupled from the file one. I have actually made attempts on decoupling it, that is kept in some private branch.

**Comment by viciious** [comment]
https://github.com/TTimo/GtkRadiant/issues/265

**Comment by dnk777** [comment]
The AAS world loading code currently treats AAS file contents as a memory dump, the only thing it does is byte-swapping. That's how the genuine Q3 code used to work, and I have kept this behavior. If file and memory representation differ, a garbage is loaded.

**Comment by viciious** [comment]
Representation is the same, it's just that the number of clusters and associated data is different. It's valid and compatible across the platforms, but different.

**Comment by dnk777** [comment]
I have modified AAS loading code a bit so it prints to console 
https://gist.github.com/dnk777/a3173397bbf4e1f6e4be0be4d11412ff

My native AAS File
```
......
Gametype 'Free for All' initialized
Num bboxes: 2
Num vertexes: 4602
Num planes: 1992
Num edges: 9505
Edge index size: 27278
Num faces: 5829
Face index size: 9372
Num areas: 1006
Num area settings: 1006
Num reach.: 1493
Num nodes: 3600
Num portals: 1
Portal index size: 0
Num clusters: 3
AiAasWorld: 29 floor clusters, 3 stairs clusters (including dummy zero ones) have been detected
....
```


The given 450K file

```
....
Gametype 'Free for All' initialized
Num bboxes: 2
Num vertexes: 3601
Num planes: 1646
Num edges: 7397
Edge index size: 20821
Num faces: 4490
Face index size: 6861
Num areas: 770
Num area settings: 770
Num reach.: 1284
Num nodes: 2430
Num portals: 1
Portal index size: 0
Num clusters: 2
AiAasWorld: 34 floor clusters, 2 stairs clusters (including dummy zero ones) have been detected
....
```

My native BSPC output
```
....
writing basewsw/maps/wda5.aas
numvertexes = 4602
numplanes = 1992
numedges = 9505
edgeindexsize = 27278
numfaces = 5829
faceindexsize = 9372
numareas = 1006
numareasettings = 1006
reachabilitysize = 1493
numnodes = 3600
numportals = 1
portalindexsize = 0
numclusters = 3
....
```

Try using that (if it does not lead to crash) http://rgho.st/8DKX2PyFf

**Comment by dnk777** [comment]
That's what is expected https://webmshare.com/q5DgO. I here avoid triggering the bot's aggression by walking and not making any sounds, so the bot roams over the map most of the time. And I have actually started addressing various actions issues only in b6057d685f0adff07872751c188061baf82ac02e, so many things are still missing, like melee attacks (that should be planned), etc. The video starts from bot picking RL near the YA spawn and only then attacking me (when the bot has picked good weapons for close range attacking). If the bot ignores RA, that's for a reason (the bot is stacked and does not have a logic for self-damaging in this case (it should an action partially satisfying some `DenyItem` goal, along with some `GoToItem` action), the bot often visits the upper glass dome area for shards which are pickable.

This code shows area boundaries near "Player" and gives an idea how crooked the AAS thing is
https://gist.github.com/dnk777/718b08ffd5f860696a2c97620c078b11

**Comment by viciious** [comment]
https://youtu.be/9LFqE_IF5tU

Your aas file seems to play better indeed but make sure you see the whole thing

**Comment by dnk777** [comment]
If you want a bot to attack more, copy `ffa.weigths` or even `rekt.weights` to `dm_wda5.weights`. They prefer roaming and picking items in DM as the weights have emerged (what goal is chosen having a plan satisfying it can be seen by enabling https://github.com/Qfusion/qfusion/blob/master/source/game/ai/ai_base_brain.cpp#L616, green = grab item, red = kill enemy, etc, these colors are defined in bot_goals.h), attacking while executing this goal is occasional (if the view angles for aiming and running match), but this action temporary suffers from https://github.com/Qfusion/qfusion/commit/cf2490dccfc2761245c7891f2fa1404f965083f4 (I have temporarily disabled it due to jitter in choice it introduces when a bot loses an enemy out of sight and forgets it, until a proper solution arrives). Due to that quick fix they sometimes try rush through enemy/enemy area ignoring it. The code for rejection of blocked goals used to exist for 1.5 years, first version is here https://github.com/Qfusion/qfusion/commit/cfad03515b3329cd5e6bfe3c8173fb3cc1cca4f0. I have only started inspecting goals/actions for real (the movement have been still hurting me a lot until the recent patch).

The bot gets blocked on stairs near YA because AAS is exceptionally bad here, and I prefer stand still if there is no solution atm (if there are no prediction results and a-priori feasible actions also fail).
https://webmshare.com/bZLvy
That's how Q3 AAS looks like (i think they put botclip brushes on stairs so they do not produce a mess)
https://webmshare.com/4B4g4

If you wish, I can revert all /ai and related scripts code to vanilla 2.1 state, create an issue and I will do it.

**Comment by viciious** [comment]
While maps can certainly be tuned to improve bot performance, it'd still be nice to know why AAS files produced by x86 and x64 versions of bspc are so radically different yet valid. Is it the arch that is to blame or the compiler?

**Comment by dnk777** [comment]
I don't know, we're lucky the AAS files at least are binary compatible across platforms (we're lucky that sizeof(int) is 4 on all platforms, and AAS files mostly use ints in definitions of data types). What we (I) can do: AAS loading code should separate in-memory and file data type representation (as I said, I tried to do it, but stopped working on it for two reasons: lack of time and in order to avoid further patch growth), and all AAS files should be built using the version of the compiler that produces best results.

I think there is an arch-dependent define somewhere in BSPC code that affects logic too (like some branches are cut off by this define)

**Comment by dnk777** [comment]
On https://github.com/Qfusion/qfusion/commit/cf2490dccfc2761245c7891f2fa1404f965083f4 and tied to it stupid bot behavior in some cases: I have decided to temporarily turn it off because the proper solution described in the comment diff should involve modification of enemy tracking, and I deferred this task for the same reasons as decoupling AAS file and in-memory representation (it is desirable not only to avoid further issues with binary compatibility, but the routing algorithm could be speed up by adding several additional fields to AAS data types, so there is no need in various lookup tables that lead to cache misses noticeable in the cachegrind tool).
